### ISSUE TITLE
Implement TaxEngine and Calculator as services

### ISSUE THREAD

**Issue by roadster31** [thelia/thelia#2424]
TaxEngine and Calculator are hardwired classes and not services, thus it is impossible to implement different tax systems

**Comment by zorn-v** [comment]
TaxEngine seems already defined and used as service.

**Comment by roadster31** [comment]
True. But not the Calculator.

**Comment by zorn-v** [comment]
1. In controllers we have container
2. In action we can inject that new service
But how to inject it to models ?

PS. And there are many typos in current Thelia\TaxEngine\Calculator (method names in TaxEngineException text)

**Comment by zorn-v** [comment]
No, you must first set dispatcher for this. And if you already must do it, there is no difference to set calculator.

**Comment by anoziere** [comment]
Status check on current code (T3 `main`, and 2.6 where relevant).

Part of this is already done. PR #3094 (merged 2023-03-20) made tax types services and made them resolvable from a Propel model:

- `core/lib/Thelia/Core/TheliaKernel.php:471` auto-tags any `TaxTypeInterface` with `thelia.taxType` (2.6: `core/lib/Thelia/Core/Thelia.php:476`).
- `core/lib/Thelia/Action/Tax.php:29-34` holds `#[AutowireLocator('thelia.taxType')]` and answers `TheliaEvents::TAX_GET_TYPE_SERVICE` at `core/lib/Thelia/Action/Tax.php:80-86`.
- `core/lib/Thelia/Model/Tax.php:52-66` obtains the dispatcher from the Propel connection and resolves the type through that event.

So a module can add a tax type today. That answers the question raised in the thread in 2017 ("how to inject it to models?"): the dispatcher is reachable from a model via `Propel::getServiceContainer()->getWriteConnection(...)->getEventDispatcher()`.

What is still hardwired is the `Calculator` itself. In T3 it is instantiated with `new Calculator()` in 11 places, four of which are Propel model methods with no container access: `core/lib/Thelia/Model/Product.php:51` and `:58`, `core/lib/Thelia/Model/ProductSaleElements.php:69` and `:79`, `core/lib/Thelia/Model/CartItem.php:182` and `:192`, `core/lib/Thelia/Model/TaxRule.php:25`. The others are `core/lib/Thelia/Action/Sale.php:117`, `core/lib/Thelia/Module/AbstractDeliveryModuleWithState.php:76`, `core/lib/Thelia/Domain/DataTransfer/Export/Type/ProductTaxedPricesExport.php:37` and the back-office themes.

Two constraints that the 2017 discussion did not cover:

- `Calculator` is stateful: `load()`, `loadTaxRule()`, `loadTaxRuleWithoutCountry()`, `loadTaxRuleWithoutProduct()` (`core/lib/Thelia/Domain/Taxation/TaxEngine/Calculator.php:150,174,202,227`) mutate the instance and return `$this`. A shared service would leak state between calls, so it would have to be non-shared or cloned per use.
- It exposes four public static methods (`Calculator.php:59,72,80,114`) used for cart and order tax factors. Those cannot be overridden through DI and would have to become instance methods.

If this is picked up, the workable shape is a `TaxCalculatorInterface` with a non-shared default service, resolved from the models through a `TAX_GET_CALCULATOR` event, i.e. the same mechanism already used by `Model\Tax::getTypeInstance()`. Existing tax modules extend tax types, not the Calculator, so they would not be affected. T3 has already moved the class from `Thelia\TaxEngine\Calculator` to `Thelia\Domain\Taxation\TaxEngine\Calculator` with no `class_alias`, so the API is already in a breaking window there; 2.6 should not change.

One data point on priority: a code search over the `thelia-modules` org returns no implementation of `TaxTypeInterface`, so the extension point delivered in 2023 has no consumer yet. Keeping this open as a T3-only item, retitled to what remains (making the Calculator replaceable), and dropping the 2.6 label.

**Comment by anoziere** [comment]
Second half shipped in https://github.com/thelia/thelia/pull/3587, merged into `main`. Leaving this open for what remains, listed at the bottom.

The calculator is now replaceable. `TaxCalculatorInterface` and `TaxCalculatorFactoryInterface`/`TaxCalculatorFactory` are reachable two ways: constructor injection where a service can be autowired (`Action\\Sale` — the only such site), and a `TAX_GET_CALCULATOR` event for the six sites that have no container, through `TaxCalculatorResolverTrait`. That is the same route `Model\\Tax::getTypeInstance()` already takes. `Action\\Tax` answers at priority `-128` and only when no listener has already provided one, so a module listener at any normal priority wins.

Why a factory rather than an injected service, since that question decides the whole design: the calculator is stateful. `load()`, `loadTaxRule()`, `loadTaxRuleWithoutCountry()` and `loadTaxRuleWithoutProduct()` all mutate `$this` and return it, and `loadTaxRuleWithoutCountry()` resets the product and the collection while leaving `country` and `state` stale. Reusing one instance across two different load kinds is therefore silently wrong. A factory gives one instance per computation, exactly matching what `new Calculator()` does today.

Backward compatibility is intact: `new Calculator()` still works, no signature changed, no constructor of a module-extended class changed, and the trait falls back to `new Calculator()` when the Propel connection carries no dispatcher (install scripts, standalone CLI). That matters — `GoogleShoppingXml`, `CarriersDelivery`, `Doofinder`, `EasyProductManager`, `CustomDelivery` and `GoogleTagManager` all instantiate it directly.

Still open here:

1. The four static methods (`getUntaxedCartDiscount`, `getUntaxedOrderDiscount`, `getCartTaxFactor`, `getOrderTaxFactor`) cannot be replaced. Making them replaceable means instance methods, which PHP will not allow next to statics of the same name, so it needs a rename or a removal — a BC decision, not something to settle inside this PR.
2. The two back-office themes still build the calculator directly, in `thelia-templates/default-twig` and `thelia-templates/back`. Separate repositories, separate changes.

Unrelated to this issue but found while mapping the call sites, and worth its own ticket: `getOrderTaxFactor()` and `getCartTaxFactor()` cache their result in a function-static variable (`Calculator.php:83` and `:117`), shared for the whole request across every order and cart. The second order handled in one request gets the first order's tax factor.

**Comment by anoziere** [comment]
Closing: the calculator is replaceable, which is what this issue asked for. The four static helpers and the two back-office themes are tracked in #3594.
